【Task28】Create Forgot Password page②

csakyo opened the pull request:
# [Task28]
パスワードを忘れた方へ画面作成
https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#28

## [codesandbox](https://codesandbox.io/s/github/csakyo/homework/tree/feature/task28-2/question28?file=/js/password.js) / Latest commit:https://github.com/csakyo/homework/pull/43/commits/76fdaed82425a2677a611f7c2694dbdc00799890

## 課題28仕様
パスワード忘れた方へ(forgetpassword.html)のページを作ってください

## 前回実装済
この課題から会員登録された内容はlocalstorage内に保有され ログイン時は固定値ではなく localstorage内の値とチェックされます。 イメージしたいのはローカルストレージはここではサーバー内のテーブルとして使っています。 なので現実的ではそのような実装はないですが、ここの課題はそれで実装したいです 

## 今回実装点
①会員登録画面(register.js)
会員登録画面でsubmitした時、localStorage内の値とチェックしてメールアドレスが既に登録されていた場合はエラーを出します。

②パスワードをお忘れの方へ(forgotpassword.js)
上記で作ったメールアドレス兼ユーザー名を入力するところがあります
メールアドレスのinput値がバリデーションを通れば送信ボタンが押せます
もしメールアドレスが登録がされていない場合は「見つかりませんでした」というエラーメッセージを出してください
この課題の実装では forgetpassword.htmlでsubmitしたらトークンを発行して(仮にこの説明では482r22fafahとします)、ローカルストレージに保存。 そのあと register/password.htmlへ飛ばしてください
その際のリンクはregister/password.html?token=482r22fafah みたいなurlにして(このリンクが本来メール本文にあるリンクのイメージです)

③パスワード再設定ページ(password.js)
「新規パスワード入力」 「確認のためのパスワード入力」 があり、バリデーションも効きます

### 確認してほしい点
【動作確認してほしい箇所】
①会員登録画面(register.html)
一度会員登録した後、ローカルストレージに既にあるメールアドレスで再登録しようとするとエラーが出る

↓パスワードをお忘れの方へページへ遷移してもらい

②パスワードをお忘れの方へページ(forgotpassword.html)
・入力値が間違っている場合エラーが出る
・メールアドレスのバリデーションが通ればsubmitボタンが押せて、password.htmlへ飛べる

↓forgotpassword.htmlから遷移すると

③パスワード再設定ページ
・パラメータ(発行されたトークン(tokenForforgotPassword))付きのURLになっている
・新規PWと確認用PWの入力値を入れればsubmitボタン押せる
・新規PWと確認用PWの入力値が相違ある場合にエラーが出る（submitボタン押せない）

### 前回の課題から変更したファイル
・register.js
・forgotpassword.js
・/module/validation.js

### 前回の課題から追加したファイル
・password.js

## 未実装箇所
フロントはurlパラメータのtokenがすぐ前で発行した482r22fafahかどうかを確認します。 (ローカルストレージがサーバー側の役割をしているイメージ)
もしtokenが間違っていたら notautherize.html 権限がありませんページに飛ばしてください
urlパラメータが正しい場合にregister/password.htmlを表示します
表示は下記の通りで

inputの値が同じでバリデーションが通ればパスワードを再発行と新たなトークンを発行して ローカルストレージに埋め込みor以前のそれを破棄、変更してください

パスワード再設定完了ページ
register/passworddone.html?token=tagaerega <-新しいトークン で飛ばします
完了ページでは urlパラメータからトークンと直近で作られたトークンがローカルストレージのそれと合っていたら
「パスワード再設定が完了しました」 「ログインページへ」
を表示し
間違っていたりトークンがなかったり直叩きでurlが間違っていたら 権限なしページに飛ばしてください
その後、新たに作ったパスワードでログインができるようにします

## [Comment for this Task]
Please see below for the changes and additions that have been made.

csakyo committed:
Feat: add to change password page and forgotpassword page

csakyo commented on question28/js/register.js:
```diff
@@ -63,7 +63,15 @@ checkbox.addEventListener("change", () => {
+    mailInputArea.nextElementSibling.textContent = "すでに登録されているメールアドレスです";
+    submitButton.disabled = true;
+    return;
+  }
```
①register.js
If the e-mail address is already registered, an error message will appear.

csakyo commented on question28/js/forgotpassword.js:
```diff
@@ -19,21 +19,44 @@ mailInputArea.addEventListener('blur', (e) => {
+const checkData = (emailAddressValue) => {
   const registeredUserData = JSON.parse(localStorage.getItem('userData'));
+  return ((emailAddressValue === registeredUserData.user_email));
+}
```
②forgotpassword.js
Compares input values with values in local storage.

csakyo commented on question28/js/forgotpassword.js:
```diff
@@ -19,21 +19,44 @@ mailInputArea.addEventListener('blur', (e) => {
+  const tokenForPasswordReset = verificationResult.token;
+  const passwordResetPageUrl = "./register/password.html";
+  localStorage.setItem('tokenForforgotPassword', tokenForPasswordReset);
+  window.location.href = `${passwordResetPageUrl}?token=${tokenForPasswordReset}`;
```
②forgotpassword.js
Emits an error if the value in the local storage does not match the input value. 
If the value in local storage matches the input value, a token is embedded and the user is moved to password.html with the token as a parameter.

csakyo commented on question28/js/password.js:
```diff
@@ -0,0 +1,35 @@
+
+const checkAllValidity = () => {
+  return document.getElementsByTagName("input").length === document.getElementsByClassName("valid").length && passwordInputArea.value === confirmPasswordInputArea.value;
+}
```
③password.js
On this page, validation passes if the input values are consistent, in addition to if the valid class is satisfied.

csakyo commented on question28/js/password.js:
```diff
@@ -0,0 +1,35 @@
+    
+    toggleDisabledOfSubmitButton(checkAllValidity());
+  });
+}
```
③password.js
When a blur event fires, validation runs for each password.(validateInputValue())
If both are entered, it checks if the entered values are correct and displays an error if there is a discrepancy.
The submit button is activated when all validations have been passed.

haru-programming reviewed (commented):
[ share ]
I am sharing this because I learned a lot from Moriken-sensei's sharing.
It is out of scope, so no corrections are needed.

> I have the impression that target.id is like a unique string like a token, but it is actually name or email. target.name (name attribute) or target.from would be more appropriate.

https://github.com/haru-programming/JavaScript-lessons/pull/34#discussion_r1032847318

haru-programming commented:
[ MUST ]
In register/password.html,
it seems to be possible to submit even if the input does not meet the 8 or more large and small alphanumeric characters.

`My input : haru1234`

Incidentally, the validation also passed on the member registration screen, also with` haru1234`.

haru-programming commented on question28/js/forgotpassword.js:
```diff
@@ -19,21 +19,44 @@ mailInputArea.addEventListener('blur', (e) => {
-  const inputMailData = mailInputArea.value;
+const checkData = (emailAddressValue) => {
   const registeredUserData = JSON.parse(localStorage.getItem('userData'));
+  return ((emailAddressValue === registeredUserData.user_email));
```
Q.
Is it okay without the ()?
```suggestion
  return emailAddressValue === registeredUserData.user_email;
```

csakyo replied:
It's true! Thanks for pointing that out! Fixed so please confirm.

fixed commit: https://github.com/csakyo/homework/pull/43/commits/a16567ab2abd5745f39951cfcc51acb526fe6763

haru-programming commented on question28/js/forgotpassword.js:
```diff
@@ -19,21 +19,44 @@ mailInputArea.addEventListener('blur', (e) => {
-  window.location.href = './notautherize.html';
+
+  const tokenForPasswordReset = verificationResult.token;
+  const passwordResetPageUrl = "./register/password.html";
```
Q.
should I make` passwordResetPageUrl` variable?
I wrote directly.
I'd really appreciate it if you could enlighten me:)!

csakyo replied:
Thank you for your review!
You are right. I thought it was not necessary to put it in a variable since it would not be used elsewhere. Fixed so please confirm!

fixed commit: https://github.com/csakyo/homework/pull/43/commits/00331682b361c303e83fa7e4196fc3da9ad58c6a

haru-programming commented on question28/js/password.js:
```diff
@@ -0,0 +1,35 @@
+
+const checkAllValidity = () => {
+  return document.getElementsByTagName("input").length === document.getElementsByClassName("valid").length && passwordInputArea.value === confirmPasswordInputArea.value;
+}
```
[ IMO ]
Since it returns a boolean value, I felt that the function name `is**` would be easier to understand.

csakyo replied:
Thank you for your review!
You are right. I changed the variable names to ``isValidAllInputsValue`` from ``checkAllValidity``. What do you think? Please advise if there is a more appropriate one. :bow:

fixed commit : https://github.com/csakyo/homework/pull/43/commits/c1ac6abb0927d113efa4d7d37940ba7df8ab5256

haru-programming commented on question28/js/register.js:
```diff
@@ -63,7 +63,15 @@ checkbox.addEventListener("change", () => {
-  const userData = Object.fromEntries([...new FormData(form)]);
-  localStorage.setItem('userData', JSON.stringify(userData)); 
+  const registeredUserData = JSON.parse(localStorage.getItem('userData'));
+  const inputUserData = Object.fromEntries([...new FormData(form)]);
```
[ comment ]
`const inputUserData = Object.fromEntries([...new FormData(form)]);`
It was very smartly written and I learned a lot!
I didn't know about new FormData().
Thank you!

nakagawamai replied:
comment
I've never heard of `new FormData ()` either! Learning, thank you!

nakagawamai commented on question28/register/password.html:
```diff
@@ -14,13 +15,13 @@ <h1>Password Reset</h1>
-                            <label for="password">確認のためのパスワード入力</label>
-                            <input type="password" id="password" name="user_password" required class="input"/>
+                            <label for="confirmPassword">確認のためのパスワード入力</label>
+                            <input type="password" id="confirmPassword" name="confirm_password" required class="input js-confirm-password"/>
```
IMO
I thought adding an autocomplete attribute would make it a better form.
ref : https://web.dev/sign-in-form-best-practices/#new-password

csakyo replied:
Thank you for your review and giving the ref!
I added autocomplete attribute to password inputs.The id name was also changed accordingly. Please confirm.

register.html / password.html →new-password
login.html→current-password

On the password.html page, I was worried about how to handle the confirmation password, but I used new-password for the autocomplete attribute and confirm-password for the id. (Because the same id cannot be used.)

fixed commit : 
https://github.com/csakyo/homework/pull/43/commits/a231b99077d39c2620f18efc8b97e5f4a72be0b3
https://github.com/csakyo/homework/pull/43/commits/76fdaed82425a2677a611f7c2694dbdc00799890

nakagawamai commented on question28/js/password.js:
```diff
@@ -0,0 +1,35 @@
+  submitButton.disabled = !isValid;
+}
+
+const showErrorMessageInputValuesMismatch = () => confirmPasswordInputArea.nextElementSibling.textContent = passwordInputArea.value !== confirmPasswordInputArea.value ? "入力されたパスワードが一致しませんでした" : "";
```
nits
It seemed a little complicated.
I thought `passwordInputArea.value ! == confirmPasswordInputArea.value` could be put together, but what do you think?
```
    if (passwordInputArea.value && confirmPasswordInputArea.value) {
      showErrorMessageInputValuesMismatch(); 
    }
```

```JavaScript
    if (passwordInputArea.value && confirmPasswordInputArea.value && passwordInputArea.value !== confirmPasswordInputArea.value) {
      // Write the process here
    }
```

csakyo replied:
Thank you for your review!
You are right. I have tried to reflect your opinion and write it in one condition. 
When there is no problem with validation, the error statement needs to be removed, so I have added that as well.
The validation here is a bit complicated and troubling part...so if you have any more appropriate ideas would be appreciated:bow:

fixed commit : https://github.com/csakyo/homework/pull/43/commits/709a5fea5cb2d2dc8d1756a6e7fc0bcbac05f0eb

csakyo committed:
Refactor: Fixed regular expression used in validation

csakyo committed:
Refactor: Remove unnecessary parentheses

csakyo committed:
Refactor: Change URLs put in variables to direct writing

csakyo committed:
Refactor: Correct variable names as appropriate.

csakyo committed:
Refactor: Added autocomplete attribute

csakyo committed:
Refactor: Rewriting if statements

csakyo committed:
Refactor: Validation specification changed to use name attribute instead of id

csakyo committed:
Refactor: Change the id name to match the autocomplete attribute

csakyo commented:
> [ share ] I am sharing this because I learned a lot from Moriken-sensei's sharing. It is out of scope, so no corrections are needed.
> 
> > I have the impression that target.id is like a unique string like a token, but it is actually name or email. target.name (name attribute) or target.from would be more appropriate.
> 

@haru-programming 
Thank you for sharing the infomation! I learned a lot as well! 
I just tried to use the name attribute instead of id. I apologize that it may be difficult to confirm this, as various changes have occurred on the pages that use validation 🙇 

fixed commit : https://github.com/csakyo/homework/pull/43/commits/356a47fd340ef0cb8f06d295ef67dd1d5c337458

csakyo commented:
> [ MUST ] In register/password.html, it seems to be possible to submit even if the input does not meet the 8 or more large and small alphanumeric characters.
> 
> `My input : haru1234`
> 
> Incidentally, the validation also passed on the member registration screen, also with` haru1234`.

@haru-programming 
Thank you for your review! I have completely overlooked this...!
I fixed validation checks on regular expressions.Please confirm.

fixed commit : https://github.com/csakyo/homework/pull/43/commits/96372f11d801536e26d467176e4560c3273d0ab6

haru-programming commented:
Thank you for the correction..!!
I thought it MUST be partially OK!

[ IMO ]
I would like to share what reviewed by Mai-san,but I don't think the password for confirmation needs to be more than 8 characters or more than 8 large and small alphanumeric validation.
